receiver/awscontainerinsightreceiver/internal/neuron/pod_attribute_decorator.go

```diff
@@ -0,0 +1,149 @@
+	for i := 0; i < metricDatapoints.Len(); i++ {
+		attributes := metricDatapoints.At(i).Attributes()
+		var containerInfo *stores.ContainerInfo
+
+		if neuronDeviceIndex, neuronDeviceIndexPresent := attributes.Get(neuronDeviceAttributeKey); neuronDeviceIndexPresent {
+			// get container info from neuronDeviceIndex
+			neuronDeviceIndex := neuronDeviceIndex.AsString()
+			containerInfo = pdc.getContainerInfoForNeuronDeviceIndex(neuronDeviceIndex)
+
+		} else if neuronCoreIndex, neuronCoreIndexPresent := attributes.Get(neuronCoreAttributeKey); neuronCoreIndexPresent {
+			// get container info from neuronCore
+			containerInfo = pdc.PodResourcesStore.GetContainerInfo(neuronCoreIndex.AsString(), neuronCoreResourceName)
+			neuronDeviceIndex := getNeuronDeviceIndexFromCoreAttribute(neuronCoreIndex, neuronCoresPerDevice)
+			if containerInfo == nil {
+				// else get container info from calculated neuronDeviceIndex
+				containerInfo = pdc.getContainerInfoForNeuronDeviceIndex(neuronDeviceIndex)
+			}
+			attributes.PutStr(neuronDeviceAttributeKey, neuronDeviceIndex)
+		}
+		populateAttributes(&attributes, containerInfo)
+	}
+}
+
+func (pdc *PodAttributesDecoratorConsumer) getContainerInfoForNeuronDeviceIndex(neuronDeviceIndex string) *stores.ContainerInfo {
+	containerInfo := pdc.PodResourcesStore.GetContainerInfo(neuronDeviceIndex, neuronDeviceResourceName)
+	if containerInfo == nil {
+		// Alt resource name is to support backward compatibility in neuron monitor : https://awsdocs-neuron.readthedocs-hosted.com/en/latest/containers/tutorials/k8s-setup.html
+		containerInfo = pdc.PodResourcesStore.GetContainerInfo(neuronDeviceIndex, neuronDeviceResourceNameAlt)
+	}
+	return containerInfo
+}
+
+func populateAttributes(attributes *pcommon.Map, containerInfo *stores.ContainerInfo) {
+	if containerInfo != nil {
+		attributes.PutStr(ci.AttributeContainerName, containerInfo.ContainerName)
+		attributes.PutStr(ci.AttributeK8sPodName, containerInfo.PodName)
+		attributes.PutStr(ci.AttributeK8sNamespace, containerInfo.Namespace)
+	}
+}
+
+func getMetricDatapoints(m pmetric.Metric) pmetric.NumberDataPointSlice {
+	switch m.Type() {
+	case pmetric.MetricTypeGauge:
+		return m.Gauge().DataPoints()
+	case pmetric.MetricTypeSum:
+		return m.Sum().DataPoints()
+	default:
+		return pmetric.NewNumberDataPointSlice()
+	}
+}
+
+// We extract the attribute named `neuroncore_per_device_count` from the metric to get the value
+// https://awsdocs-neuron.readthedocs-hosted.com/en/latest/tools/neuron-sys-tools/neuron-monitor-user-guide.html
+func getNeuronCoresPerDevice(neuronHardwareInfo pmetric.Metric) int {
+	neuronCoresPerDeviceValue, _ := neuronHardwareInfo.Sum().DataPoints().At(0).Attributes().Get(neuronCorePerDeviceKey)
+	neuronCoresPerDevice, _ := strconv.Atoi(neuronCoresPerDeviceValue.AsString())
+	return neuronCoresPerDevice
```

Comment:
It is safe to assume this will always return a value > 0? just wondering if we need to check anything beforehand just in case like `.At(0)` is nil or something in the chain?

Reply:
Well the NeuronHardware Metric if found should always be a Sum type Pmetric and in an ideal scenario should always have datapoints > 0, but I will add appropriate error handling in this case